### PR TITLE
Add support to secure APIs using mutual SSL

### PR DESCRIPTION
**Pull request by Megala21**
## Purpose
Currently APIM do not have support to securing APIs using mutual SSL. With this PR, that support will be added.

## Goals
This PR fixes https://github.com/wso2/product-apim/issues/3819. Add out-of-box support for securing APIs using mutual SSL.

## Approach
In the UI, during API creation process in the manage phase, API creator will can select OAuth2 or Mutual SSL or both as follow,
screenshot1

The client certificates can be uploaded using following UI. While uploading the client certificate, relevant subscription tier can be selected.
screenshot2

Uploaded certificates will be listed like below,
screenshot3

From store, relevant API can be invoked as below,
screenshot4

Same thing can be achieved using REST APIs as well. Please refer the swagger docs for the details on newly introduced APIs.

## User stories
https://docs.google.com/document/d/1syUw22Re9wLbomyYfQAP-EI-UWl9FnBrCGLHJ0L54Kg/edit?usp=gmail

## Release note
Support for securing APIs using mutual SSL

## Documentation
https://docs.google.com/document/d/1jCSQdwa1MJqNVwGkSSVUgdswu4lP1flRxCd1kXZbmjg/edit?usp=sharing

## Training
N/A

## Marketing
N/A

## Automation tests
Unit and integration test added.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
N/A

## Related PRs
https://github.com/wso2/product-apim/pull/3820

## Migrations (if applicable)
To use this feature, new table need to be created.

## Test environment
Ubuntu 18.04, JDK 1.8
 
## Learning
How Mutual SSL works

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java [hunk 1]
```diff
@@ -901,22 +902,53 @@ boolean updateAPIforStateChange(APIIdentifier identifier, APIStatus newStatus,
...
      */
     int deleteCertificate(String userName, String alias, String endpoint) throws APIManagementException;
 
+    /**
+     * Method to remove the client certificates which is mapped to given alias and api identifier from database.
+     *
+     * @param userName      : Name of the logged in user.
+     * @param apiIdentifier : Identifier of API for which the certificate need to be deleted.
+     * @param alias         : Alias of the certificate which needs to be deleted.
+     * @return Status that indicate whether delete client certificate succeeds.
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java]: What are the possible values here?

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java [hunk 2]
```diff
@@ -954,6 +1007,17 @@ boolean updateAPIforStateChange(APIIdentifier identifier, APIStatus newStatus,
      */
     boolean isCertificatePresent(int tenantId, String alias) throws APIManagementException;
 
+    /**
+     * Method to check whether an client certificate for the given alis is present in trust store and whether it can
+     * be modified by current user.
+     *
+     * @param tenantId Id of the tenant.
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java]: missing colon

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java [hunk 3]
```diff
@@ -974,6 +1038,20 @@ boolean updateAPIforStateChange(APIIdentifier identifier, APIStatus newStatus,
      */
     int updateCertificate(String certificateString, String alias) throws APIManagementException;
 
+    /**
+     * Method to update the existing client certificate.
+     *
+     * @param certificate   : Relevant certificate that need to be updated.
+     * @param alias         : Alias of the certificate.
+     * @param APIIdentifier API Identifier of the certificate.
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java]: missing colon

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/dto/ClientCertificateDTO.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     * @param tierName Name of the tier.
+     */
+    public void setTierName(String tierName) {
+        this.tierName = tierName;
+    }
+
+    /**
+     * To get the alias of the certificate.
+     *
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/dto/ClientCertificateDTO.java]: do we need this extra line?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityException;
+import org.wso2.carbon.apimgt.gateway.handlers.security.Authenticator;
+import org.wso2.carbon.apimgt.gateway.handlers.security.oauth.OAuthAuthenticator;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+
+/**
+ * Authenticator responsible for handling Mutual SSL and OAuth secured APIs.
+ */
+public class MutualSSLAndOAuthAuthenticator implements Authenticator {
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: In the case of mutual ssl and oauth can't we use the OAuth and MutualSSL authenticators one after the other rather than having separate authenticator?

**Comment by harsha89** [comment]
@Megala21 can we have some screenshots attached to the feature?

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                } else {
+                    return i18n.t("Failed to delete Certificate due to an Internal Server Error.");
+                }
+            }
+            case (3) : {
+                return i18n.t("Failed to add Certificate in to Publisher Trust Store. Certificate exists for" +
+                    " the Alias");
+            }
+            case (4) : {
+                return i18n.t("Failed to delete the certificate. Certificate could not found for the given alias." +
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js]: Shall we rephrase this as "Failed to delete the certificate. Certificate could not **be** found for the given alias. " The highlighted word is missing,

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+                    " Hence the entry is removed from the data base.");
+            }
+            case (6) : {
+                return i18n.t("Failed to add Certificate. Certificate Expired");
+            }
+            case (7) : {
+                return i18n.t("Failed to add Certificate. Certificate for Endpoint exists.");
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js]: 'Endpoint' shouldn't have capital 'E'.Same with the word 'Certificate' when used in the middle of a sentence

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+             * */
+            var aliasMatched = existingCertificates.filter(function (certificate) {
+                return certificate.alias === alias;
+            });
+
+
+            if (aliasMatched.length > 0) {
+                jagg.message({
+                    type: "error",
+                    content: i18n.t("Could not add certificate for alias") + ", '" + alias + "'. " +
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js]: Good message. Let's add the step what should be done to overcome this as well.

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            var apiName = $(e.currentTarget).attr("data-apiname");
+            var apiVersion = $(e.currentTarget).attr("data-apiversion");
+            var apiProvider = $(e.currentTarget).attr("data-apiprovider");
+            var certUi = this;
+
+            jagg.message({
+                content: deleteConfirmation + " " + dataAlias + "? <br/> <strong>" +
+                i18n.t("This action cannot be undone") + ".</strong>",
+                type: "confirm",
+                title: i18n.t("Delete Certificate with alias") + " " + dataAlias,
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js]: only proper nouns should start with a capital letter

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+            $("#messageModal div.modal-footer").html("");
+            var dataAlias = $(e.currentTarget).attr("data-alias");
+            var apiName = $(e.currentTarget).attr("data-apiname");
+            var apiVersion = $(e.currentTarget).attr("data-apiversion");
+            var apiProvider = $(e.currentTarget).attr("data-apiprovider");
+            var certUi = this;
+
+            jagg.message({
+                content: deleteConfirmation + " " + dataAlias + "? <br/> <strong>" +
+                i18n.t("This action cannot be undone") + ".</strong>",
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-manage/js/jquery.certUi.js]: Shall we reword it and say, This action cannot be reversed?

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json [hunk 11]
```diff
@@ -650,6 +656,11 @@
...
+    "You have selected Microgateways for this API. The mediation policies selected here will not be applied in Microgateway environments.":"You have selected Microgateways for this API. The mediation policies selected here will not be applied in Microgateway environments.",
+    "This option determines the type of security that will be used to secure this API. An API can be secured with either one of them or it can be secured by both of them. If OAuth2 option is selected, relevant API will require a valid OAuth2 token for successful invocation. If Mutual SSL option is selected, a trusted client certificate should be presented to access the API":"This option determines the type of security that will be used to secure this API. An API can be secured with either one of them or it can be secured by both of them. If OAuth2 option is selected, relevant API will require a valid OAuth2 token for successful invocation. If Mutual SSL option is selected, a trusted client certificate should be presented to access the API",
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json]: Could you tell me where this message is used so that I can help you rephrase this to bring out the meaning better? Can you attach a UI screen shot?

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json [hunk 12]
```diff
@@ -526,6 +531,7 @@
     "Upload API definition as": "Upload API definition as",
     "Upload Certificate": "Upload Certificate",
     "Upload Endpoint Certificates": "Upload Endpoint Certificates",
+    "Upload client certificates" : "Upload client certificates",
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json]: Can we add multiple certs at the same time? Can you attach a UI screen shot?

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json [hunk 13]
```diff
@@ -379,6 +382,7 @@
     "Password": "Password",
     "Please select at least one subscription tier.": "Please select at least one subscription tier.",
     "Please select at least one transport.": "Please select at least one transport.",
+    "Please select at least one API Security.": "Please select at least one API Security.",
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json]: shouldn't it be at least one API Security method? I am not sure the context this is used in, can you share a UI screenshot to understand better?

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json [hunk 14]
```diff
@@ -342,6 +344,7 @@
     "No documentation associated with the API": "No documentation associated with the API",
     "No matching APIs found.": "No matching APIs found.",
     "No uploaded certificates for the endpoints": "No uploaded certificates for the endpoints",
+    "No uploaded client certificates for the API": "No uploaded client certificates for the API",
```
Comment (shaniR) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json]: "No client certificates were found uploaded for this API." or its even better if we can just say "No client certificate found for this API. Please upload a certificate"

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+            }
+            authenticator = oAuthAuthenticator;
+        } else {
+            if (mutualSSLAuthenticator == null) {
+                mutualSSLAuthenticator = new MutualSSLAuthenticator(apiLevelPolicy, certificateInformation);
+            }
+            authenticator = mutualSSLAuthenticator;
+        }
+        return authenticator;
+    }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: I think this if-else branch has no use. why can't we simply instantiate and init 2 authenticators at line 60-61?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+            String[] splitParts = apiWithversion.split(":");
+            api = splitParts[0];
+            apiWithversion = splitParts[1].substring(1);
+        }
+        return new APIIdentifier(apiPublisher, api, apiWithversion);
+    }
+
+    @Override
+    public String getChallengeString() {
+        return "Mutual SSL realm=\"WSO2 API Manager\"";
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: What is this used for? Having "WSO2" hard-coded can be a security issue in case this is exposed to clients.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java [hunk 17]
```diff
@@ -81,6 +145,27 @@ public ResponseCode addCertificateToParentNode(String certificate, String alias,
...
+            String tierName, int tenantId) {
+        ResponseCode responseCode;
+        try {
+            responseCode = certificateMgtUtils.validateCertificate(alias, tenantId, certificate);
+            if (responseCode == ResponseCode.SUCCESS && certificateMgtDAO
+                    .checkWhetherAliasExist(alias, tenantId)) {
+                responseCode = ResponseCode.ALIAS_EXISTS_IN_TRUST_STORE;
+            }
+            if (responseCode == ResponseCode.SUCCESS) {
+                certificateMgtDAO.addClientCertificate(certificate, apiIdentifier, alias, tierName, tenantId);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java]: so we add the cert even if it exists? is that intentional?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java [hunk 18]
```diff
@@ -81,6 +145,27 @@ public ResponseCode addCertificateToParentNode(String certificate, String alias,
...
+            }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java]: we don't have to check for `responsecode=success` twice if we make line 154 a nested if.  

```
if(success){
   if (exists){
     error
   } esle { //depends on the answer to the above question
   addcert()
  }
} 
```

**Comment by Megala21** [comment]
@harsha89 - I have added the screenshots to the description.. Please have a look. Thanks.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java (hunk 5 above)
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: We can use that. But when the user invokes the API every time, we need to go through an if else construct, to identify the api security for the relevant API. But if we have a separate authentictaor, the selection of authenticator is one time task only when publishing and we do not need to go through the check every time we invoke the API. WDYT?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java (hunk 16 above)
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: When there is an API invocation failure due to authorization failure, we add a http header "WWW-Authenticate" with this value from authenticator. For our OAuthAuthenticator, we have "OAuth2 realm=\"WSO2 API Manager\"", I followed the same pattern.  If we are to change this, shall we add "Mutual SSL" only?

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/dto/ClientCertificateDTO.java (hunk 4 above)
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/dto/ClientCertificateDTO.java]: This is the default format for java doc method level comment. It will be added automatically when we format it.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java [hunk 19]
```diff
@@ -533,6 +551,112 @@ public String getAPIEndpointSecurityType(API api, String tenantDomain) throws AP
...
+            return;
+        }
+        try {
+            CertificateMgtDAO.getInstance().updateRemovedCertificatesFromGateways(api.getId(),
+                    APIUtil.getTenantIdFromTenantDomain(tenantDomain));
+            /* The flow does not need to be blocked, as this failure do not related with updating client certificates
+             in gateway, rather updating in database. There is no harm in database having outdated certificate
+             information.*/
+        } catch (CertificateManagementException e) {
+            log.error("Certificate Management Exception while trying to update the remove certificate from gateways "
```
[the remark this answers is not in the document]
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java]: Thanks for the suggestion. In the case, when we are logging an error log due to an exception, we normally add stack trace as the second parameter in log. Stacktrace will include all the information on what happened and how to solve it. For example, in this case, if this CertificateManagementException has happened due a foreign key violation in database, stacktrace will include the relevant error code from the database level and then the server admin knows this is the issue that is happening and they know the path for the rectification.
Further in this place, we catch a more generic exception, which is wrapping a API Management Exception or SQL Exception/ or other lower level exception. So I think we cannot know the exact reason for the error at this point, either solution. Only way for the user to figure out them is from stack trace.WDYT?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CertificateMgtUtils.java [hunk 20]
```diff
@@ -141,6 +160,64 @@ public ResponseCode addCertificateToTrustStore(String base64Cert, String alias)
...
+        } catch (IOException e) {
+            log.error("I/O Exception while trying to load trust store while trying to check whether alias " + alias
+                    + " exists", e);
+            responseCode = ResponseCode.INTERNAL_SERVER_ERROR;
+        } catch (CertificateException e) {
+            log.error("Certificate Exception while trying to load trust store while trying to check whether alias "
+                    + alias + " exists", e);
+            responseCode = ResponseCode.INTERNAL_SERVER_ERROR;
+        } catch (NoSuchAlgorithmException e) {
+            log.error("No Such Algorithm Exception while trying to load trust store while trying to check whether "
```
[the remark this answers is not in the document]
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CertificateMgtUtils.java]: Thanks for the suggestion on this. We have catched the NoSuchAlgorithmException at this point and in the error message we are try to tell the user on at which scenario this exception has occured. Regarding why has it occured and how to overcome this error, I think for that User can refer the oracle documentation on "NoSuchAlgorithmException" at [1] https://docs.oracle.com/javase/8/docs/api/java/security/NoSuchAlgorithmException.html. If we are trying to do include the same information here, it will be more of a duplicate. WDYT?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java (hunk 5 above)
Comment (bhathiya) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: Even now, doesn't it go through this in each and every call? 

https://github.com/wso2/carbon-apimgt/blob/a379b19cf21a9b031b5767733bd95c8852308dcc/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java#L202

Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: Basically i am suggesting something similar below.
```
         if (isOAuthProtected) {
                authenticator = new OAuthAuthenticator(authorizationHeader, removeOAuthHeadersFromOutMessage);
authenticator.authenticate()
            } 
            if(isMutualSSLProtected) {
                authenticator = new MutualSSLAuthenticator(apiLevelPolicy, certificateInformation);
authenticator.authenticate()
            }
if(authenticator == null) {
// Give some error
}
```

OR

authenticator could be a list of authenticators

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java (hunk 16 above)
Comment (bhathiya) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: Shall we read the product name from the carbon.xml and use it here?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java (hunk 5 above)
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: @bhathiya  - It does not go through this particular segment every time, only when initializing the authentication handler, as I checked we go through the particular code segment only if authenticatior is null. Hence it will only be executed once in server startup time and when the relevant API is published again.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java (hunk 16 above)
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: Sure.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java (hunk 5 above)
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: @isharac - During the implementation phase, I also thought of both possibilities (i.e. Having list of authenticators in APIAuthenticationHandler and separate authenticator for combined authentication) and finally decided to go ahead with this design due to the reason I explained before and as well as considering following, we have recommended the customers to use the APIAuthenticationHandler as the extension point, if they want to go with the custom authentication method. In that case if we are introduce the list of authenticators,  For instance if the customer is extending and using this APIAuthenticationHandler, we cannot change the getAuthenticator method(with return type authenticator) as it is protected and the customer may be overriding it to initialize his authenticator and further we use that particular method when there is authentication failure. Hence to make sure not to break the backward compatibility I made sure to include a new authenticator. If this is an issue from maintainable perspective, for example if we are introduce a new type of API security mechanism, we may need to go with multiple authenticators (To include 2 secure mechanism, to include 3 secure mechanism). To avoid this shall I rename this as MultiAuthenticator, so that we will be able to use the same authenticator safely if we are to use different authentication mechanism in future.

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java (hunk 2 above)
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java]: AFAIK colon is not a mandatory in java doc comment. However since we have followed the same pattern in this interface in the some other methods, I will make changes to include the colon only in this particular interface.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java (hunk 5 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: If the problem is back word compatibility, in the case of multiple authenticators what if we can have another method which will return list of enabled authenticators? And getAuthenticator() will return null.

So in the case of getAuthenticator()==null we can look for getAuthenticators()

At authentication failure we use this authenticator to get ChallengeString. In the case of  MutualSSLAndOAuthAuthenticator  you do a string concatenation to get this.
so we can do this in the APIAuthenticationHandler itself by looking at the list of authenticators.
wdyt?

Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java]: As discussed offline, we will be having a separate authenticator for multiple authentication support and will be changing the class in more generic way to support additional protocols.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/ClientCertificatesApi.java [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+    @Path("/{alias}/content")
+    @Consumes({ "application/json" })
+    @Produces({ "application/json" })
+    @io.swagger.annotations.ApiOperation(value = "Download a certificate.", notes = "This operation can be used to download a certificate which matches the given alias.\n", response = void.class)
+    @io.swagger.annotations.ApiResponses(value = { 
+        @io.swagger.annotations.ApiResponse(code = 200, message = "OK.\n"),
+        
+        @io.swagger.annotations.ApiResponse(code = 400, message = "Bad Request.\nInvalid request or validation error.\n*\n"),
+        
+        @io.swagger.annotations.ApiResponse(code = 404, message = "Not Found. Certificate for the Alias not found.\n"),
```
[the remark this answers is not in the document]
Comment (Megala21) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/ClientCertificatesApi.java]: We are following the same pattern in other APIs as well. Hence I think we can have it

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json (hunk 11 above)
Comment (Megala21) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json]: This message is used as the help text, when the user is trying the select the API security.
screenshot from 2018-10-12 17-10-54

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json (hunk 12 above)
Comment (Megala21) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json]: No. We cannot add it . Now I have modified this. Attached updated UI
screenshot from 2018-10-12 17-11-08

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json (hunk 13 above)
Comment (Megala21) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/conf/locales/jaggery/locale_default.json]: Yes. That should be API security method. I have attached the modified UI.
screenshot from 2018-10-12 17-14-07

**Comment by Megala21** [comment]
@bhathiya , @isharac @shaniR - I have fixed most of the review comments and responded to some. Please have a look. Thanks.
